crates/protobuf_extractor/src/extract_protobuf_imports.rs

```diff
@@ -0,0 +1,56 @@
+use anyhow::Result;
+use tree_sitter::Parser;
+
+#[derive(Debug, PartialEq)]
+pub struct ProtobufSource {
+    pub imports: Vec<String>
+}
+
+pub fn parse(source: &str, _source_path: &str) -> Result<ProtobufSource> {
+    let mut buf = Vec::default();
+    let mut parser = Parser::new();
+    parser.set_language(tree_sitter_proto::language()).expect("tree-sitter failed to initialize");
+    let tree = parser.parse(source, None).expect("parse failed");
+    let root_node = tree.root_node();
+    let mut cursor = root_node.walk();
+    let bytes = source.as_bytes();
+    for child_node in root_node.children(&mut cursor) {
+        match child_node.kind() {
+            "import" =>
+                {
+                    for path in child_node.children_by_field_name("path", &mut child_node.walk()) {
+                        let mut quoted = path.utf8_text(bytes).unwrap().chars();
+                        // Remove the quotation marks
```

Comment:
This is interesting in the level of parsing vs not in this. -- the lack of quotes extraction. Could you try extend this next maybe to try get a set of the messages/services defined in the file? curious how easy/hard or possibly error prone or not that is. (i guess how much is left to the app here vs the tree sitter handles for us i'm curious to see)

Reply:
if we are only parsing the imports, I wonder if we are winning that much by using tree-sitter (which *seems* to me pretty low level still).

On the other hand, we probably eventually want all the message names and the package as well very soon, so doing all that with nom may be a pain. But seems like maybe close (and we would also have an easier time adding features for linting, etc...)

Reply:
tree-sitter on its own is mostly a parser generator with some runtime. The actual grammar comes from mitchellh/tree-sitter-proto, which defined the `import` tree as https://github.com/mitchellh/tree-sitter-proto/blob/42d82fa18f8afe59b5fc0b16c207ee4f84cb185f/grammar.js#L34-L39

```js
    import: $ => seq(
      'import',
      optional(choice('weak', 'public')),
      field('path', $.string),
      ';',
    ),
```

In my code `path.utf8_text(bytes)` basically extracts the substring of what's marked as `'path'`, which in this case is a `$.string` node, which is completely unaware of quotation because each language does slightly different things with it (single quote, triple quote, string interpolation etc).

Reply:
The benefit of tree-sitter-based parsers is that the run-time is usually much faster than hand parsing with regex because the state machine has been expanded out in C code, also it supports partial parsing and ignore the parts that it doesn't understand. Eventually I think this could be used for Scala parsing as well.

We also have to parse comments if we want to support directives.